[Chee Jia Le] iP

jiale-c opened the pull request:
## Mum
> “It is what it is.” – Chee Jia Le [(source)](https://github.com/jiale-c/) :shipit:

Mum frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ **_SUPER FAST_** to use

 All you need to do is,

1. download it from [here.](https://github.com/jiale-c/)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
###
```JAVA
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jiale-c committed:
Add greet, echo and exit.

jiale-c committed:
Revert "Add greet, echo and exit."

This reverts commit f73e253ed94493e70cc451b55e8371d4bd617a90.

jiale-c committed:
Added Greet, Echo, Exit.

jiale-c committed:
Added greet, echo and exit

jiale-c committed:
Add add, list.

jiale-c committed:
Add mark as done

jiale-c committed:
Add ToDos, Events, Deadlines

jiale-c committed:
Add automated text ui testing.

jiale-c committed:
Add handle errors.

jiale-c committed:
Add delete.

jiale-c committed:
Add enums.

jiale-c committed:
Add Save

jiale-c committed:
Add dates and times format for deadline and event.

jiale-c committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add Save

jiale-c committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add dates and times format for deadline and event.

jiale-c committed:
Add bug fix to allow data to be stored in file and load from file when running Duke.

jiale-c committed:
Add more OOP

jiale-c committed:
Add organisation into packages

jiale-c committed:
Add JUnit Tests

jiale-c committed:
Add JavaDoc comments

jiale-c committed:
Add following the coding standard

jiale-c committed:
Add find command

jiale-c committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

jiale-c committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add following the coding standard

# Conflicts:
#	src/main/java/duke/commands/Command.java
#	src/main/java/duke/storage/Storage.java

jiale-c committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find command

jiale-c committed:
Add find class into parse class

jiale-c committed:
Merge branch 'add-gradle-support' of https://github.com/jiale-c/ip

kaiyichen commented on src/main/java/duke/commands/ShowAllCommands.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Enumeration of all valid commands.
+     */
+    enum Commands {LIST, TODO, EVENT, DEADLINE, BYE, MARK, UNMARK, DELETE, FIND}
```
I like how you put the commands as enum which makes your following code neater :thumbsup:

kaiyichen commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+ * @param <T>
+ */
+public abstract class Command<T> {
+    private void execute() throws DukeException, DukeException {}
```
Any reason why you did not make it a public abstract method to be overridden by subclasses since all subclasses implements execute() ?

kaiyichen commented on src/main/java/duke/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+ * Command class that adds a deadline task to list.
+ */
+public class AddDeadlineCommand extends Command<String> {
+    private TaskList list;
```
I like how you use non-public variables :smiley:

kaiyichen commented on src/main/java/duke/commands/ShowAllCommands.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    private void execute() {
+        System.out.print(Ui.lineDivider + "COMMANDS:\n");
+        switch (Commands.LIST) {
+            case LIST:
+                System.out.println("list    | Get a list of tasks you have");
+            case TODO:
+                System.out.println("todo    | One of the 3 tasks type. eg: todo_<yourtask>");
+            case EVENT:
+                System.out.println("event   | One of the 3 tasks type. eg: event_<yourtask>_/at_<YYYY-MM-DD HH:MM>");
+            case DEADLINE:
+                System.out.println("deadline| One of the 3 tasks type. eg: deadline_<yourtask>_/by_<YYYY-MM-DD HH:MM>");
+            case BYE:
+                System.out.println("bye     | End this phone call");
+            case UNMARK:
+                System.out.println("unmark  | Unmark the task as not done. eg: unmark_<task number>");
+            case MARK:
+                System.out.println("mark    | Mark the task as done. eg: mark_<task number>");
+            case DELETE:
+                System.out.println("delete  | Delete the task from list. eg: delete_<task number>");
+            case FIND:
+                System.out.println("find    | Find the tasks with a keyword from list. eg: find_<keyword>");
```
Should there be no indentation for case clauses? :thinking:

kaiyichen reviewed (commented):
Overall code is very neat and nicely written. It follows the coding standard well too. Just some nits to fix.

jiale-c committed:
Add Gradle testing files

jiale-c committed:
Add gradle and its testing files

jiale-c committed:
Add checkstyle and used checkstyle in java files

junjunjieOng commented on src/main/java/duke/commands/MarkCommand.java:
```diff
@@ -0,0 +1,83 @@
+package duke.commands;
+
+import duke.commands.Command;
```
Hi👋 , should this import be with line 7 to 11?

junjunjieOng commented on src/main/java/duke/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+ */
+public class AddDeadlineCommand extends Command<String> {
+    private TaskList list;
+    private String[] echo;
```
Perhaps we can change this String[] name into plural form? 🤔

junjunjieOng commented on src/main/java/duke/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,69 @@
+
+    private void execute() throws DukeException {
+        String err = "Oh no! The description of event cannot be empty... Try again :)\n";
+        String wrongFormat = "Oh no! The format for event task is wrong... Try again :)\n";
```
I like how you declare and initialise the message first, so that is easier to change the message if need to in the future. 😄

junjunjieOng commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    /**
+     * Obtain answer on whether this task is or is not completed.
```
Should this doc sentence change to follow the coding standard?
``` 
     Obtains answer on whether this task is or is not completed.
```

junjunjieOng reviewed (commented):
Overall code looks neat and awesome. Just some minor changes to fix.

jiale-c committed:
Add GUI to duke

jiale-c committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI to duke

jiale-c committed:
Add a proper exit of application when input bye

jiale-c committed:
Add some edits to ShowAllCommands to look better

jiale-c committed:
Add assertion to command types

Assertions help define assumptions about the commands added so that
runtime can verify them.

Asserts are added throughout the classes to ensure that the format of
user input is correct during runtime.

The sooner the execution stops, it is safer.

Refer to CS2103 website to find out more for the use of assertions
https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/errorHandling.html#assertions

jiale-c committed:
Add assertion to tasks in list and storage

Assertions help define assumptions about the task added so that
runtime can verify them.

Asserts are added throughout the classes to ensure that the format of
user input is correct during runtime.

The sooner the execution stops, it is safer.

Refer to CS2103 website to find out more for the use of assertions
https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/errorHandling.html#assertions

jiale-c committed:
Clean up code

Unused variables and imports were present.

Unused variables causes more overhead and trip up reader.

Code looks neater when there are no old code that serves no purpose but
just confuses the reader.

jiale-c committed:
Add SLAP practises and default branch in Storage

Code with multiple deep nesting makes it look messy.

Deep nesting in the storage file can cause code duplication and making
it prone to errors. Applying SLAP makes the code look neater without
arrowheads. Switching to have default case to detect errors.

Using SLAP is preferable over deep nesting in this situation because
common behaviours are not separated.

Refer to CS2103 notes on dealing with code quality
https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/
codeQuality.html#code-quality

jiale-c committed:
Add Streams

jiale-c committed:
Merge pull request #1 from jiale-c/branch-A-Assertions

branch-A-Assertions

jiale-c committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertion to tasks in list and storage
  Add assertion to command types

jiale-c committed:
Merge pull request #2 from jiale-c/branch-A-CodeQuality

branch-A-CodeQuality

jiale-c committed:
Merge branch 'master' into branch-A-Streams

* master:
  Add SLAP practises and default branch in Storage
  Clean up code
  Add assertion to tasks in list and storage
  Add assertion to command types

# Conflicts:
#	src/main/java/duke/commands/FindCommand.java

jiale-c committed:
Merge pull request #3 from jiale-c/branch-A-Streams

branch-A-Streams

jiale-c committed:
Add a places command

Additional functionality to include places in event task.

Have a choice to include a location into event. Bot can list out a list
of specific places if user intend to find all tasks at that location.

Having PlacesCommand so that command can vet if a task in list contains
the specific task.

jiale-c committed:
Apply SLAP to execute() of commands

The execute command can be further SLAP upon as it is too long.

Extract out code to separate methods to make it more readable.

jiale-c committed:
Edit javadoc of duke class

jiale-c committed:
Add SLAP to storage

jiale-c committed:
Remove redundant exceptions throwing

jiale-c committed:
Edit commands to allow assert to work

jiale-c committed:
Add css styling for gui

jiale-c committed:
Add readme.md to describe the app to new users

jiale-c committed:
Edit user guide

jiale-c committed:
Merge branch 'branch-D-Places'

* branch-D-Places:
  Add a places command

jiale-c committed:
Merge D-Places branch to main branch

jiale-c committed:
Add places to readme and command summary list in ShowAllCommands class

jiale-c committed:
Start with an empty list when first time using mum

jiale-c committed:
Fix typos in user guide

jiale-c committed:
Update README.md

jiale-c committed:
Add minor fix up to storage class

jiale-c committed:
Add minor fix up for storage class

jiale-c committed:
Merge branch 'master' of https://github.com/jiale-c/ip

* 'master' of https://github.com/jiale-c/ip:
  Update README.md
